### PR TITLE
Move snapshotting files to own package

### PR DESCRIPTION
**Pull request by wolfs**
Snapshotting is now in `org.gradle.internal.snapshot`. Moreover, I renamed `NormalizedFileSnapshot` to `FileFingerprint`.

The only thing used by the `snapshot` package which is still in `org.gradle.api.internal.changedetection.state` is `WellKnownFileLocations`. What could be a good location for that class (and its implementation)? I am tempted to put it in `org.gradle.internal.file` or `org.gradle.internal.hash` in `baseServices`. Another option would be to put it in `org.gradle.cache` in `persistentCache`. WDYT?

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/OverlappingOutputs.java [hunk 1]
```diff
@@ -18,8 +18,8 @@
 
 import org.gradle.internal.file.FileType;
 import org.gradle.internal.fingerprint.FileCollectionFingerprint;
+import org.gradle.internal.fingerprint.FileFingerprint;
 import org.gradle.internal.fingerprint.HistoricalFileCollectionFingerprint;
-import org.gradle.internal.fingerprint.NormalizedFileSnapshot;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/OverlappingOutputs.java]: I like this change!

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java [hunk 2]
```diff
@@ -19,6 +19,7 @@
 import org.gradle.cache.PersistentCache;
 import org.gradle.cache.PersistentIndexedCache;
 import org.gradle.cache.PersistentIndexedCacheParameters;
+import org.gradle.internal.snapshot.FileSystemSnapshotter;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java]: Is this right?

**Comment by wolfs** [comment]
> I don't think this has much to do with persistentCache, so let's not put it there. Also, org.gradle.internal.hash doesn't sound related: that's a small utility package that deals with hashing in a pretty low-level sense. org.gradle.internal.file sounds like it has some connected things, but since the API of WellKnownFileLocations refers to CachedJarFileStore, we should either abstract that away a bit first, or just put the whole thing into org.gradle.internal.classpath, since that's what this is (currently) used for anyway.

It is not used for classpath, but for any hashes which we calculate - file hashes of transforms and dependencies would be part of that, two. I think the Gradle API hash is also handled by that.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/DefaultTaskOutputFilesRepository.java]: Well spotted! Now that we can determine the roots from the output file fingerprints, the whole declared output file data is not necessary anymore. Moreover, `FileSystemSnapshotter.exists` can go away and we never hit the file system when updating the generated output files generation! Deleting all that code felt so good...

**Comment by lptr** [comment]
Whoa, we just ended up deleting a lot of stuff! That looks great, but I'd like to understand what happened here. Let's discuss!

**Comment by wolfs** [comment]
I thought a little bit more about your naming suggestions. `FileSystemObjectSnapshot` doesn't sound so bad after all. The only thing I dislike is that the name is so long...

Are you against `RegularFileSnapshot`? I think `FileSnapshot` would not be enough given that we also have `MissingFileSnapshot`, right? Except if we call the base class `FileSnapshot` (currently `PhysicalSnapshot`).

I think `PhysicalSnapshot` -> `FileSnapshot` isn't so bad after all, is it?

**Pull request merged**